src/v/cluster/client_quota_store.h

```diff
@@ -0,0 +1,75 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "client_quota_serde.h"
+#include "container/fragmented_vector.h"
+
+#include <absl/algorithm/container.h>
+#include <absl/container/node_hash_map.h>
+
+namespace cluster::client_quota {
+
+class store final {
+public:
+    using container_type = absl::node_hash_map<entity_key, entity_value>;
+    using range_containter_type
+      = fragmented_vector<std::pair<entity_key, entity_value>>;
+
+    /// Upserts the given quota at the given entity key
+    /// All quota types are overwritten with the given entity_value, so on alter
+    /// operations we need to read the current state of the quota and merge it
+    /// with the alterations
+    void set_quota(const entity_key&, const std::optional<entity_value>&);
```

Comment:
question: Is there any significance to providing a `std::nullopt` for the second parameter?

Reply:
An alternative implementation might call `set_quota(key, std::nullopt)` as `remove_quota(key)`.

Reply:
Yeah I'd say it isn't clear via the documentation.... so maybe either a specific "remove" or document that std::nullopt means remove.